[DataPipe] Separating DataPipes from Dataset into different files

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73991
* __->__ #73396

Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

NivekT committed:
[DataPipe] Separating DataPipes from Dataset into different files

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34481962).

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

ejguan reviewed (commented):
Overall looks great! Thank you

Have you tested it with `import torchdata`. Just want to verify if this PR introduces any BC breaking changes.

And, what do you think about the name/location of this `datapipe.py` file because we have a folder called `datapipes` at the same directory. It can be confusing for users when doing `import` in python. Is that possible to move this file into `datapipes` folder but expose the API at the same place.

ejguan commented on torch/utils/data/datapipe.py:
```diff
@@ -0,0 +1,199 @@
+
+from torch.utils.data._typing import _DataPipeMeta
+from torch.utils.data._utils.serialization import serialize_fn, SerializationType, deserialize_fn
+from torch.utils.data.dataset import IterableDataset, T_co, UNTRACABLE_DATAFRAME_PIPES, Dataset
```
`T_co` and `UNTRACABLE_DATAFRAME_PIPES` should be created from this file rather than importing them.

And, if we do `import torch.utils.data.dataset.IterableDataset` and `import torch.utils.data.dataset.Dataset`, would circular import still happen?

NivekT replied:
I just refactored it, though most of the import statements for DataPipes have to be changed.

I am explicitly leaving `graph.py` and `graph_settings.py` in `torch/utils/data` (instead of moving them into the `datapipes` directory) because they are needed by `DataLoader`. Let me know if you think those two files should be moved as well.

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34481962).

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -17,19 +14,12 @@
-                              ]
 
 class DataChunk(list, Generic[T]):
     def __init__(self, items):
```
Is that possible to move `DataChunk` out of `dataset.py` as well?

NivekT replied:
Yep, I just did and updated this PR.

ejguan reviewed (approved):
Thank you soooo much! This is looks awesome.

NivekT committed:
Update on "[DataPipe] Separating DataPipes from Dataset into different files"


Separating DataPipes from Dataset into different files. This makes the code more maintainable and simplifies some of the code generation.

I have also tried to move `datapipe.py` into `torch.utils.data.datapipes`, but that will lead to circular import and rewriting many import statements. Should I put more time and go down that path some more?

Fixes https://github.com/pytorch/data/issues/213

Differential Revision: [D34481962](https://our.internmc.facebook.com/intern/diff/D34481962)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34481962).
